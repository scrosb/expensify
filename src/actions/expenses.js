import uuid from 'uuid';
import database from '../firebase/firebase';

//component calls action generator
//action generator returns object
//component dispatches object
//redux store changes.

//component calls action generator
//action generator returns function
//component dispatches function
//function runs

//YOU MUST PASS IN A OBJECT TO A FUNCTION THAT TAKES AN OBJECT, YOU WON'T GET ANY ERRORS OTHERWISE


//ADD_EXPENSE
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {

    const uid = getState().auth.uid;
    const {
      description = '', 
      note = '', 
      amount=0, 
      createdAt = 0 
    } = expenseData;

    const expense = {
      description,
      note, 
      amount,
      createdAt
    }

    //save expense to the database then dispatch the action causing a redirect
    return database.ref(`users/${uid}/expenses`).push(expense).then((ref) => {
      dispatch(addExpense({
        id:ref.key,
        ...expense
      }))
    })
  };
}


//REMOVE_EXPENSE takes an OBJECT!!!!!!

export const removeExpense = ({id} = {}) => ({
  type:'REMOVE_EXPENSE',
  id
});

//Remove from state and database

export const startRemoveExpense = ({id} = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/expenses/${id}`).remove().then(() => {
      dispatch(removeExpense({ id }));
    });
  }
}



//EDIT_EXPENSE

export const editExpense = (id, updates) => ({
  type:'EDIT_EXPENSE',
  id,
  updates
});


//startEditExpense

export const startEditExpense = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses/${id}`).update(updates).then(() => {
      dispatch(editExpense(id, updates));
    })
    // database.ref().update({
    //   'job/company':'Amazon',
    //   stressLevel:9,
    //   'location/city':'Seattle'
    // }).then(() => {
    //   console.log('company, city, stress updated')
    // }).catch((e) => {
    //   console.log('Errors: ', e);
    // })
  }
}

//SET_EXPENSES

export const setExpenses = (expenses) => ({
  type:'SET_EXPENSES',
  expenses
})

//this will fetch the data, then call setExpenses

export const startSetExpenses = () => {
  return (dispatch, getState) => {

    const uid = getState().auth.uid;

    //return a promise
    return database.ref(`users/${uid}/expenses`).once('value').then((snapshot) => {
          const expenses = [];
          snapshot.forEach((childSnapshot) => {
            expenses.push({
              id:childSnapshot.key,
              ...childSnapshot.val()
            })
          })
          //add expenses to the state
          dispatch(setExpenses(expenses));
    });
  };
};
