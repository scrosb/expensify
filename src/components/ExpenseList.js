import React from 'react';
//connect imports the store to you component
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

//add export here for snapshot testing
export const ExpenseList = (props) => {
  return (
    <div>
      {
        props.expenses.length === 0 ? (
          <p>No Expenses</p>
        ) : (
          props.expenses.map((expense) => 
          (
          <ExpenseListItem
          key={expense.id}
          {...expense}
          />
          )
        )
        )
      }
    </div>
)};


//passing application state to props using connect
//as the store changes this will automatically rerender
const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
}


export default connect(mapStateToProps)(ExpenseList);


