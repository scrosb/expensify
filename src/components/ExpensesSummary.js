import React from 'react';
import {connect} from 'react-redux';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';


// numeral(amount / 100).format('$0,0.00')
export class ExpensesSummary extends React.Component {
  //expense count
  //expenses total
  render(){
    return (
      <div>
        {this.props.expenses.length=== 0 ? 'No Expenses Found':
        <p>Viewing {this.props.expenses.length} expense(s) totaling &nbsp;
         { numeral(selectExpensesTotal(this.props.expenses)/100).format('$0,0.00')}</p>}
      </div>
    );
    }
}

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
}

export default connect(mapStateToProps)(ExpensesSummary);
