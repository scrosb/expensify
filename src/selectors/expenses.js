//Get Visible Expenses
import moment from 'moment'

export default (expenses, {text, sortBy, startDate, endDate}) => {
  return expenses.filter((expense) => {
    const createdAtMoment = moment(expense.createdAt);
    //true if not a number, only if start date is a number do we want to filter
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day'): true ;
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day'):true
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

    //figure out if expenses.description as the text variable string inside of it
    //includes
    //convert both strings to lowercase

    return startDateMatch && endDateMatch && textMatch;
  }).sort((a,b) => {
    if(sortBy === 'date'){
      //sort by created At
      return a.createdAt < b.createdAt ? 1 : -1
    }

    if(sortBy === 'amount'){
      return a.amount < b.amount ? 1:-1;
    }
  })
}
//Store Creatio