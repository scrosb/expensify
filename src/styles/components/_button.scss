.button {
  padding: $s-size;
  color:white;
  background-color: $blue;
  display: inline-block;
  border:none;
  font-weight:300;
  line-height: 1;
  font-size: $font-size-lg;
  text-decoration: none;
}

.button--link{
  @extend .button;
  background: none;
  text-decoration:none;
  color: white;
  border: none;
}

.button--secondary{
  @extend .button;
  background: #888888;
}