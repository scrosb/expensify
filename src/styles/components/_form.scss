.form {
  display:flex;
  flex-direction: column;
  > * {
    //targets all descendants of form
    margin-bottom: $m-size;
  }
}

.form__error{
  margin: 0 0 $m-size 0;
  font-style: italic;

}