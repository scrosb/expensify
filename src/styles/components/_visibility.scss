.show-for-mobile {
  //only show on mobile
  @media (min-width: $desktop-breakpoint) {
    display: none;
  }
}

.show-for-desktop {
  @media (max-width: $desktop-breakpoint - .1rem) {
    display: none;
  }
}