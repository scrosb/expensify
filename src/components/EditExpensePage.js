import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {editExpense, removeExpense} from '../actions/expenses';

//refactor edit expense to be a class based component
//setup mapDispatchToProps, Editexpense and remove expense

//should render editexpense page
//snapshot

//should handle editExpense
//spies

//should handle removeexpense
//spies
export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push('/')
  }

  onRemove = () => {
    this.props.removeExpense({id: this.props.expense.id});
    this.props.history.push('/')
  }

  render(){
    return (
      <div>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmit}
        />
        <button onClick={this.onRemove}>Remove</button>
      </div>
    )
  }
}

//Remove expense via dispatch then redirect to dashboard pae

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find((expense) => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
  editExpense: (id, expense) => dispatch(editExpense(id,expense)),
  removeExpense: (data) => dispatch(removeExpense(data))
})



export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);