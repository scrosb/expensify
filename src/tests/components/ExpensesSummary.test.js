import React from 'react';
import {shallow} from 'enzyme';
import {ExpensesSummary} from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

test('should render ExpensesSummary', () => {
  const wrapper = shallow(<ExpensesSummary expenses={expenses}/>)

  expect(wrapper).toMatchSnapshot();
})

test('should render expense summary with no expenses', () => {
  const wrapper = shallow(<ExpensesSummary expenses={[]}/>)

  expect(wrapper).toMatchSnapshot();
})

test('should render expense summary with 1 expense', () => {
  const wrapper = shallow(<ExpensesSummary expenses={[expenses[1]]}/>)

  expect(wrapper).toMatchSnapshot();
})