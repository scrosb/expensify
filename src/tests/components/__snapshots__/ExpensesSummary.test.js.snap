// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary 1`] = `
<div>
  <p>
    Viewing 
    3
     expense(s) totaling  
    $1,141.95
  </p>
</div>
`;

exports[`should render expense summary with 1 expense 1`] = `
<div>
  <p>
    Viewing 
    1
     expense(s) totaling  
    $1,095.00
  </p>
</div>
`;

exports[`should render expense summary with no expenses 1`] = `
<div>
  No Expenses Found
</div>
`;
